contracts/src/Teleporter/upgrades/TeleporterRegistry.sol

```diff
@@ -5,28 +5,109 @@
 
 pragma solidity 0.8.18;
 
-import "../../WarpProtocolRegistry.sol";
 import "../ITeleporterMessenger.sol";
+import "@subnet-evm-contracts/interfaces/IWarpMessenger.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
 
 /**
- * @dev TeleporterRegistry contract is a {WarpProtocolRegistry} and provides an upgrade
- * mechanism for {ITeleporterMessenger} contracts.
+ * @dev Registry entry that represents a mapping between protocolAddress and version.
  */
-contract TeleporterRegistry is WarpProtocolRegistry {
-    constructor(
-        ProtocolRegistryEntry[] memory initialEntries
-    ) WarpProtocolRegistry(initialEntries) {}
+struct ProtocolRegistryEntry {
+    uint256 version;
+    address protocolAddress;
+}
+
+/**
+ * @dev TeleporterRegistry contract provides an upgrade mechanism for {ITeleporterMessenger} contracts
+ * through Warp off-chain messages
+ */
+contract TeleporterRegistry {
+    // Address that the off-chain Warp message sets as the "source" address.
+    // The address is not owned by any EOA or smart contract account, so it
+    // cannot possibly be the source address of any other Warp message emitted by the VM.
+    address public constant VALIDATORS_SOURCE_ADDRESS = address(0);
+
+    WarpMessenger public constant WARP_MESSENGER =
+        WarpMessenger(0x0200000000000000000000000000000000000005);
+
+    bytes32 public immutable blockchainID;
+
+    // The latest protocol version. 0 means no protocol version has been added, and isn't a valid version.
+    uint256 public latestVersion;
+
+    // Mappings that keep track of the protocol version and corresponding contract address.
+    mapping(uint256 => address) internal _versionToAddress;
+    mapping(address => uint256) internal _addressToVersion;
```

Comment:
Also per the audit, let's name the map key value parameters

Reply:
Also should we make these private since `TeleporterRegistry` isn't intended to be inherited from?

Reply:
updated to private added key value naming